numpy/__init__.pyi

```diff
@@ -1,8 +1,91 @@
 # very simple, just enough to start running tests
+#
+import builtins
+from typing import Any, Mapping, List, Optional, Tuple, Union
 
-class ndarray: pass
+from numpy.core._internal import _ctypes
+
+class dtype:
+    alignment: int
+    byteorder: str
+    char: str
+    descr: List[Tuple[str, str]]
+    fields: Optional[Mapping[str, Union[Tuple[dtype, int], Tuple[dtype, int, str]]]]
+    flags: int
+    hasobject: bool
+    isbuiltin: int
+    isnative: bool
+    isalignedstruct: bool
+    itemsize: int
+    kind: str
+    name: str
+    names: Optional[Tuple[str, ...]]
+    num: int
+    shape: Tuple[int, ...]
+    ndim: int
+    subdtype: Optional[Tuple[dtype, Tuple[int, ...]]]
+
+
+    def newbyteoder(self, new_order:str = ...): dtype
+
+    str: builtins.str
+    type: builtins.type
+
+    @property
+    def base(self) -> dtype: ...
```

Comment:
Is there a reason why this is a property vs. an attribute for everything else? Does mypy care about the difference?

Reply:
Properties can't be assigned to unless they have a setter. With this stub, if you had `x: dtype`, mypy would allow `x.type = some_type` but not `x.base = some_dtype`.

Reply:
OK, in that case we should also convert most of the other attributes into readonly properties.